lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/evaluator/EvaluatorShim.java

```diff
@@ -0,0 +1,215 @@
+        EvaluatorShimProtocol.EvaluatorShimStatusProto
+            .newBuilder()
+            .setRemoteIdentifier(this.remoteManager.getMyIdentifier())
+            .setContainerId(this.containerId)
+            .setStatus(EvaluatorShimProtocol.EvaluatorShimStatus.ONLINE)
+            .build());
+  }
+
+  private void onStop() {
+    LOG.log(Level.FINER, "Entering EvaluatorShim.onStop().");
+
+    if (this.evaluatorProcess != null) {
+      try {
+        LOG.log(Level.INFO, "Waiting for the evaluator process to exit.");
+        this.evaluatorProcess.waitFor();
+
+        LOG.log(Level.INFO, "Destroying the Evaluator Process.");
+        this.evaluatorProcess.destroy();
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    LOG.log(Level.INFO, "Shutting down the evaluatorLaunchExecutorService.");
+    this.evaluatorLaunchExecutorService.shutdown();
+
+    try {
+      LOG.log(Level.INFO, "Closing EvaluatorShim Control channel.");
+      this.evaluatorShimCommandChannel.close();
+    } catch (Exception e) {
+      LOG.log(Level.SEVERE, "An unexpected exception occurred while attempting to close the EvaluatorShim " +
+          "control channel.");
+      throw new RuntimeException(e);
+    }
+
+    try {
+      LOG.log(Level.INFO, "Closing the Remote Manager.");
+      this.remoteManager.close();
+    } catch (Exception e) {
+      LOG.log(Level.SEVERE, "Failed to close the RemoteManager with the following exception: {0}.", e);
+    }
+
+    LOG.log(Level.INFO, "Exiting EvaluatorShim.onStop().");
+  }
+
+  private void onEvaluatorLaunch(final String launchCommand, final String evaluatorConfigString) {
+    LOG.log(Level.FINEST, "Entering EvaluatorShim.onEvaluatorLaunch().");
+
+    File evaluatorConfigurationFile = new File("local/evaluator.conf");
+    LOG.log(Level.FINER, "Persisting evaluator config at: {0}", evaluatorConfigurationFile.getAbsolutePath());
+
+    try {
+      boolean newFileCreated = evaluatorConfigurationFile.createNewFile();
+      LOG.log(Level.FINE,
+          newFileCreated ? "Created a new file for persisting evaluator configuration at {0}."
+              : "Using existing file for persisting evaluator configuration at {0}.",
+          evaluatorConfigurationFile.getAbsolutePath());
+
+      Configuration evaluatorConfiguration = this.configurationSerializer.fromString(evaluatorConfigString);
+      this.configurationSerializer.toFile(evaluatorConfiguration, evaluatorConfigurationFile);
+    } catch (final IOException | BindException e) {
+      LOG.log(Level.SEVERE, "An unexpected exception occurred while attempting to deserialize and write " +
+          "Evaluator configuration file. {0}", e);
+      throw new RuntimeException("Unable to write configuration.", e);
+    }
+
+    LOG.log(Level.INFO, "Launching the evaluator by invoking the following command: " + launchCommand);
+
+    this.evaluatorLaunchExecutorService.submit(new Thread() {
```

Comment:
why?

Reply:
I noticed this is triggered by remote message, so it is possible to be called multiple times.

When one thread goes in to a infinite loop by mistake, or by accident it took too long to finish execution, I think it will just be blocking to following tasks. 

If this only happends in theory not in this code, I think it is fine then.

Reply:
Well, this should only be called once, because the Evaluator process is maintained between REEF tasks (and it can run arbitrarily long). That being said, I wonder if it makes sense to ensure that onEvaluatorLaunch is only called once... I'll need check what the delivery guaranties of the transport protocol are.